feat(node): dependencies upgrade workflow

iliapolo opened the pull request:
Dependency upgrades in projen are currently managed by dependabot, which creates a PR for each package upgrade separately. This poses two issues:

- Creates clutter and possibly high maintenance load if PR's cannot be automatically merged.
- Prevents updates of packages that need to be upgraded in unity (for example `@aw-cdk`).

This PR introduces an alternative by using a GitHub workflow that executes `npm-check-updates` to upgrade `package.json` files, and `yarn` to synchronize the lock file.

Another problem with the current setup is that automatic merges are accomplished by two distinct mergify rules. 

- `auto-merge` label + build passes CI.
- PR's from dependabot + build passes CI.

This approach breaks down if branch protection rules require a PR approval. Since this is the best practice we want to encourage, we need to come up with a different merge mechanism. 

With this PR, dependency update PR's will be automatically approved by a different github workflow (`auto-approve.yml`).
This means we can remove the aforementioned mergify rules, and use the standard rule of merging approved PR's that pass CI build.

The auto approval workflow is enabled by default since it can utilize the default github token available to workflows, without requiring an additional projen specific token.

Example PR: https://github.com/iliapolo/constructs/pull/4

---
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

iliapolo commented on src/projen-upgrade.ts:
```diff
@@ -2,30 +2,21 @@ import { NodeProject } from './node-project';
-   *
-   * @default - auto-upgrade is disabled
-   */
-  readonly autoUpgradeSecret?: string;
```
Instead, I exposed `projenSecret` on `NodeProject` and reused it.

iliapolo commented on src/node-project.ts:
```diff
@@ -1085,3 +1045,73 @@ export interface NodeWorkflowSteps {
+   *
+   * @param options The options.
+   */
+  public static githubActions(options: AutoUpgradeDependenciesOptions = {}): DependenciesUpgrade {
+    return new DependenciesUpgrade('github-actions', options);
+  }
+
+  /**
+   * Use Dependabot (with custom options) to upgrade dependencies.
+   *
+   * @param options The options.
+   */
+  public static dependabot(options: DependabotOptions = {}): DependenciesUpgrade {
```
Not in love with this API, happy to accept alternatives :)

iliapolo commented on src/node-project.ts:
```diff
@@ -1085,3 +1045,73 @@ export interface NodeWorkflowSteps {
+  /**
+   * Use dependabot (with the default options) to upgrade dependencies.
+   */
+  public static readonly DEPENDABOT: DependenciesUpgrade = new DependenciesUpgrade('dependabot');
+
+  /**
+   * Use GitHub actions (with the default options) to upgrade dependencies.
+   */
+  public static readonly GITHUB_ACTIONS: DependenciesUpgrade = new DependenciesUpgrade('github-actions');
```
Just some sugar to provide an enum like experience for the common use-case.

eladb commented on .github/workflows/auto-approve.yml:
```diff
@@ -0,0 +1,17 @@
+      - unlabeled
+jobs:
+  approve:
+    if: contains(github.event.pull_request.labels.*.name, 'pr/auto-approve')
```
I don't see a huge value in the `pr/` prefix here.

```suggestion
    if: contains(github.event.pull_request.labels.*.name, 'auto-approve')
```

eladb commented on .github/workflows/auto-approve.yml:
```diff
@@ -0,0 +1,17 @@
+      - unlabeled
+jobs:
+  approve:
+    if: contains(github.event.pull_request.labels.*.name, 'pr/auto-approve')
```
Are we at all concerned that this basically allows maintainers to approve their own PRs even if "required for administrators" is set. @RomainMuller will likely have a strong opinion here.

iliapolo replied:
See https://github.com/projen/projen/pull/652#discussion_r603120017

eladb commented on src/__tests__/node-project.test.ts:
```diff
@@ -1,13 +1,46 @@
 
+describe('dependencies upgrade', () => {
+
+  test('defaults to github actions when a projen secret is defined', () => {
```
```suggestion
  test('defaults to a github workflow when a projen secret is defined', () => {
```

eladb commented on src/github/auto-approve.ts:
```diff
@@ -0,0 +1,54 @@
+  /**
+   * PR's assigned with this label will be auto-approved.
+   *
+   * @default pr/auto-approve
```
Defaults should be JSON encoded (and also, please change to `auto-approve`)

```suggestion
   * @default "auto-approve"
```

eladb commented on src/github/auto-approve.ts:
```diff
@@ -0,0 +1,54 @@
+      throw new Error('GitHub must be configured to enable auto approvals');
+    }
+
+    this.label = options.label ?? 'pr/auto-approve';
```
```suggestion
    this.label = options.label ?? 'auto-approve';
```

eladb commented on src/github/auto-approve.ts:
```diff
@@ -0,0 +1,54 @@
+    });
+    autoApprove.addJobs({
+      approve: {
+        'if': `contains(github.event.pull_request.labels.*.name, '${this.label}')`,
```
Mmm.. curious - what will cause this it fail if this is the same user?

eladb commented on src/github/auto-upgrade-dependencies.ts:
```diff
@@ -0,0 +1,143 @@
+  /**
+   * List of package names to ignore during the upgrade.
+   *
+   * @default - Only projen itself is ignored.
```
```suggestion
   * @default - `projen` is ignored since there is a separate workflow to upgrade it.
```

Which makes me wonder - with the combination of this and the self-mutation feature, do we really need a separate upgrade workflow for projen? Feel like we can just use this workflow to upgrade projen itself, and let the self-mutation workflow update the branch if there are any synth changes, no?

pgollucci replied:
tough question.

eladb commented on src/github/auto-approve.ts:
```diff
@@ -0,0 +1,54 @@
+/**
+ * Configures a GitHub workflow that auto-approves PR's that are assigned with a specific label.
+ */
+export class AutoApprove extends Component {
```
I am wondering how this will work with the self-mutation workflow (the current version and the future version which will submit a PR against your branch). 

I am thinking that maybe initially we can avoid auto-approve in the dependency upgrade workflows, so that if there are any synth changes that result in the update, a human will need to review them. WDYT?

eladb commented on src/github/auto-upgrade-dependencies.ts:
```diff
@@ -0,0 +1,143 @@
+/**
+ * Options for `AutoUpgradeDependencies`.
+ */
+export interface AutoUpgradeDependenciesOptions {
```
Rename to `UpgradeDependencies` instead of `AutoUpgradeDependencies` - the "auto" part is optional.

eladb commented on src/github/auto-upgrade-dependencies.ts:
```diff
@@ -0,0 +1,143 @@
+  /**
+   * At 00:00.
+   */
+  DAILY = 'daily',
```
NEVER is missing

eladb commented on src/github/auto-upgrade-dependencies.ts:
```diff
@@ -0,0 +1,143 @@
+
+    const ignore = ['projen', ...(options.ignore ?? [])];
+
+    const dependencies = project.github.addWorkflow('auto-upgrade-dependencies');
```
```suggestion
    const dependencies = project.github.addWorkflow('upgrade-dependencies');
```

eladb commented on src/github/auto-upgrade-dependencies.ts:
```diff
@@ -0,0 +1,143 @@
+          },
+          {
+            name: 'Upgrade yarn.lock dependencies',
+            run: `yarn install && yarn upgrade --pattern !(${ignore.join('|')})`,
```
We should use `npm` if the project is configured to use npm, no?

pgollucci replied:
https://github.com/projen/projen/issues/651

eladb commented on src/github/auto-upgrade-dependencies.ts:
```diff
@@ -0,0 +1,143 @@
+            name: 'Create Pull Request',
+            uses: 'peter-evans/create-pull-request@v3',
+            with: {
+              'token': `\${{ secrets.${project.projenSecret} }}`,
```
Is this really needed? I suspect a scheduled task should have a read/write token by default.

eladb reviewed (changes requested):
High level comment is whether we really need the projen upgrade workflow any more.

pgollucci commented on src/__tests__/integration/cdk-watchful.projenrc.js:
```diff
@@ -17,7 +17,7 @@ const project = new ConstructLibraryAws({
 
   // creates PRs for projen upgrades
-  projenUpgradeSecret: 'PROJEN_GITHUB_TOKEN',
+  projenSecret: 'PROJEN_GITHUB_TOKEN',
```
Why did this get renamed again ?

pgollucci commented on src/github/index.ts:
```diff
@@ -1,8 +1,10 @@
 export * from './mergify';
-export * from './auto-merge';
\ No newline at end of file
+export * from './workflows';
```
Make a workflows dir ?
workflows/auto-merge
worksflow/foo
worksflows/....
?

eladb commented:
@iliapolo any chance we can get this merged? It will quickly get stale and hard to maintain this branch...

RomainMuller commented on src/github/auto-upgrade-dependencies.ts:
```diff
@@ -0,0 +1,143 @@
+/**
+ * How often to check for new versions and raise pull requests for version updates.
+ */
+export enum AutoUpgradeDependenciesSchedule {
```
I don't like how all of these options are at midnight UTC. You'd usually want to randomise that a bit to spread the load over (free) CI environments... As a basic courtesy.

At least, there should be an option to provide one's own POSIX cron expression.

iliapolo replied:
Exposed a way to provide a custom cron expression 👍

RomainMuller commented on src/github/auto-upgrade-dependencies.ts:
```diff
@@ -0,0 +1,143 @@
+          ...project.installWorkflowSteps,
+          {
+            name: 'Upgrade package.json dependencies',
+            run: `npx npm-check-updates --upgrade --reject '${ignore.join(',')}'`,
```
Suggest making the upgrade target explicit (valid values are: `'minor' | 'patch' | 'latest'`).

```suggestion
            run: `npx npm-check-updates --upgrade --target 'minor' --reject '${ignore.join(',')}'`,
```

1. If the environment has `npm@^7`, this will actually not work, as it'll hang on a prompt asking if it's okay to install `npm-check-updates` before proceeding
2. Since you control the project settings, why not add `npm-check-updates` to the `devDependencies` of the project, and open with a dependency install stage? This has the added benefit of allowing the customer to control what version of `npm-check-updates` is being used (you may get nasty surprises otherwise, as `npx` will just get whatever the `latest` is... it may be incompatible with your invocations!
3. There are missing features here:
   1. I need to be able to major-pin certain dependencies (typically, I'd want this workflow NOT to upgrade any of my non-`dev` dependencies to a new major line -- and also `@types/node` should remain on the lowest supported `node` major). I'd be tempted to make this `pinToMajor = ['@types/node']`, perhaps with a way to opt out of the `dependencies` security.
   2. I need to be able to minor-pin certain dependencies (some packages do not honour SemVer, and so you wanna make sure they only get patch-upgraded... for example: `typescript`). I'd be tempted to make this `pinToMinor=['typescript']`
   3. This is not `yarn workspaces`/mono-repos aware. This may be fine (does `projen` support this use-case anyway?)

iliapolo replied:
- This isn't mono-repo aware, projen doesn't really support it anyway. 

- Regarding specific minor-pins - yes, feature needed. I'll add in a subsequent PR. This currently only addresses the naive case. 

- `npm-check-updates` now added to `devDependencies` 👍

RomainMuller commented on src/github/auto-upgrade-dependencies.ts:
```diff
@@ -0,0 +1,143 @@
+            run: `npx npm-check-updates --upgrade --reject '${ignore.join(',')}'`,
+          },
+          {
+            name: 'Upgrade yarn.lock dependencies',
+            run: `yarn install && yarn upgrade --pattern !(${ignore.join('|')})`,
```
This is precisely why I'd have named this `yarn-upgrade`, and would have checked that the project uses `yarn` before configuring.

iliapolo replied:
Added support for all package managers

RomainMuller commented on src/github/auto-upgrade-dependencies.ts:
```diff
@@ -0,0 +1,143 @@
+            name: 'Create Pull Request',
+            uses: 'peter-evans/create-pull-request@v3',
+            with: {
+              'token': `\${{ secrets.${project.projenSecret} }}`,
```
Check out the (relatively new) [`permissions`](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idpermissions) configuration...

Since this workflow triggers either on a `cron` or `workflow_dispatch`, it can be seen as only touching code that has been accepted into the default branch already (i.e: is "trusted")... You could grant the workflow extended permissions so you no longer have to provision a PAT for this kind of use-case.

iliapolo replied:
Yeap - using the default token with `write-all` permissions.

RomainMuller commented on src/github/auto-upgrade-dependencies.ts:
```diff
@@ -0,0 +1,143 @@
+            uses: 'peter-evans/create-pull-request@v3',
+            with: {
+              'token': `\${{ secrets.${project.projenSecret} }}`,
+              'commit-message': 'Upgrade dependencies',
+              'branch': 'github-actions/dependencies',
+              'labels': labels.join(','),
+              'title': 'chore(deps): Upgrade dependencies',
+              'body': 'Upgrade dependencies\n\n------\n\n*Automatically created by projen via GitHubActions*',
```
Those should be made user-configurable (`labels` obviously is already).

I would also not name the default branch `github-actions/dependencies` -- it's way too generic and could likely end up being shared by several dependency-management actions, which would not be good... At the very least, you'll want this to include the name of the workflow.

iliapolo replied:
Yeap, added workflow name.

RomainMuller commented on src/github/auto-upgrade-dependencies.ts:
```diff
@@ -0,0 +1,143 @@
+    switch (schedule) {
+      case AutoUpgradeDependenciesSchedule.DAILY:
+        return '0 0 * * *';
+      case AutoUpgradeDependenciesSchedule.WEEKDAY:
```
If you have `WEEKDAY` you also need `WEEKEND`. And then again... Are these defined "universally"? I'd be tempted to say a weekend is Saturday & Sunday... But folks in Israel may be tempted to say it's Friday & Saturday... Depending on *why* you want to refer to the week-end here (you likely mean "days when nobody is in the office").

This looks like trying too hard to me, and I'd kinda feel like you might want to allow people to craft their own CRON expression, possibly with factories for the "basic" cases such as daily, weekly, and monthly.

iliapolo replied:
Added custom cron expression support - this should suffice for now.

RomainMuller commented on src/node-project.ts:
```diff
@@ -1085,3 +1067,73 @@ export interface NodeWorkflowSteps {
+      case 'github-actions':
+        this.bindGitHubActions(project);
+        break;
+      default:
+        throw new Error(`Unsupported dependency upgrade type: ${this.type}`);
```
❤️

RomainMuller commented on src/node-project.ts:
```diff
@@ -1085,3 +1067,73 @@ export interface NodeWorkflowSteps {
+    }
+  }
+
+  private bindDependabot(project: NodeProject) {
+    project.github?.addDependabot((this.options ?? {}) as DependabotOptions);
+  }
+
+  private bindGitHubActions(project: NodeProject) {
+    new AutoUpgradeDependencies(project, (this.options ?? {}) as AutoUpgradeDependenciesOptions);
+  }
```
I'm not entirely sure these are necessary, as they are 2/3rd boilerplate... The one-liner content could probably have been directly in the `switch` within `bind`... But since they're already broken out... I guess it's not bad!

iliapolo replied:
Simplified by having the static factories pass a `binder` function

eladb reviewed (commented):
Looking forward for this!

eladb commented on src/upgrade-dependencies.ts:
```diff
@@ -0,0 +1,272 @@
+      const createPR: any = {
+        'needs': 'upgrade-dependencies',
+        'name': 'Create Pull Request',
+        'permissions': 'write-all',
```
Can we scope this down?

iliapolo replied:
I think so, let me try.

iliapolo replied:
Yes, scoped down to: 

- `content` (push branch)
- `pull-request` (create pr)
- `checks` (update status check - if needed)

eladb commented on src/upgrade-dependencies.ts:
```diff
@@ -0,0 +1,272 @@
+
+      if (project.buildWorkflowJobId) {
+        createPR.steps.push({
+          name: 'Update status check',
```
Can this be reused with the node build workflow somehow?

iliapolo replied:
With a little work, yes - but i'll skip it for now.

eladb commented on src/upgrade-dependencies.ts:
```diff
@@ -0,0 +1,272 @@
+                + '}'
+              + '}\'',
+          env: {
+            GITHUB_TOKEN: '${{ secrets.GITHUB_TOKEN }}',
```
Is this needed?

iliapolo replied:
Yes - its used in the authentication header of the `curl` command.

eladb commented on src/node-project.ts:
```diff
@@ -1182,3 +1201,46 @@ export interface NodeWorkflowSteps {
+  /**
+   * Upgrade via github actions.
+   */
+  public static githubActions(options: UpgradeDependenciesOptions = {}) {
```
The name `githubActions` is a bit ambiguous. It could be interpreted as "only upgrade GitHub actions". It's also not very accurate because the actual tool is not actions.

Maybe the name of the class could be different- something like `DependencyUpgradeMechanism` and then instead of githubActions we can use `ncu`?

iliapolo replied:
I've been struggling with this :\ 

`ncu` also feels funky cause its an implementation detail - if we switch to `npm update` would we have to change the name? 

I'll change to `DependencyUpgradeMechanism.githubWorkflow` and let me know what you think.

eladb commented on src/upgrade-dependencies.ts:
```diff
@@ -0,0 +1,368 @@
+  readonly schedule?: UpgradeDependenciesSchedule;
+
+  /**
+   * Which secret to use when creating the PR.
```
Describe the difference between using the default token and a user supplied token

iliapolo replied:
Added

eladb commented on src/upgrade-dependencies.ts:
```diff
@@ -0,0 +1,368 @@
+  /**
+   * Labels to apply on the PR.
+   *
+   * @default - no labels.
```
```suggestion
   * @default [] no labels.
```

eladb commented on src/upgrade-dependencies.ts:
```diff
@@ -0,0 +1,368 @@
+  readonly labels?: string[];
+
+  /**
+   * Execute 'build' after the upgrade.
```
Add some details on what that means

iliapolo replied:
Added

eladb commented on src/upgrade-dependencies.ts:
```diff
@@ -0,0 +1,368 @@
+  readonly include?: string[];
+
+  /**
+   * Include a github workflow for creating PR's that upgrades the
```
What happens if this is `false`? Does that imply dependabot? Does that imply that upgrades can be done manually?

iliapolo replied:
It simply creates a local projen task and upgrades can be done manually by running `yarn upgrade-dependencies`. It's not related to dependabot. Added to comment.

eladb commented on src/upgrade-dependencies.ts:
```diff
@@ -0,0 +1,368 @@
+   * The name of the task that will be created.
+   * This will also be the workflow name.
+   *
+   * @default 'upgrade-dependencies'.
```
```suggestion
   * @default "upgrade-dependencies"
```

Defaults should be JSON

eladb commented on src/upgrade-dependencies.ts:
```diff
@@ -0,0 +1,368 @@
+   *
+   * @default true
+   */
+  readonly ignoreProjen?: boolean;
```
Why not just make this the default for 'exclude'?

iliapolo replied:
So that when a user explicitly sets `exclude` - they are not implicitly including `projen`. This seemed simpler, and is consistent with how the dependabot component configures it.

eladb commented on src/upgrade-dependencies.ts:
```diff
@@ -0,0 +1,368 @@
+/**
+ * How often to check for new versions and raise pull requests for version upgrades.
+ */
+export class UpgradeDependenciesSchedule {
```
Style: I prefer to put all these auxiliary types _after_ the main options and class.

iliapolo replied:
👍

eladb commented on .projen/tasks.json:
```diff
@@ -133,19 +133,6 @@
         }
       ]
     },
-    "projen:upgrade": {
```
I expected to see the new task here

iliapolo replied:
No - since projen itself doesn't need a `projen:upgrade` task..unless i'm missing something - this never belonged here.

eladb replied:
I meant the "upgrade-dependencies" task

eladb commented on src/node-project.ts:
```diff
@@ -752,25 +758,38 @@ export class NodeProject extends Project {
+    const dependenciesUpgrade = options.depsUpgrade ?? defaultDependenciesUpgrade;
+    dependenciesUpgrade.bind(this);
+
+    if (dependenciesUpgrade.ignoresProjen && this.package.packageName !== 'projen') {
```
Why is this only used if ignoreProjen is true?

iliapolo replied:
Because otherwise projen is already being taken care of. Note the code below only configures a projen upgrade.

eladb commented on src/upgrade-dependencies.ts:
```diff
@@ -0,0 +1,368 @@
+
+  private createUpgrade(task: Task): Upgrade {
+
+    const project = this.project as NodeProject;
```
I'd keep `nodeProject` as a member (otherwise this can break in the future)

iliapolo replied:
👍

eladb reviewed (approved):
See some comments. Added `do-not-merge`

eladb reviewed (changes requested):
No do not merge so changed review verdict to avoid auto-merge

iliapolo commented on src/github/auto-merge.ts:
```diff
@@ -33,7 +25,6 @@ export interface AutoMergeOptions {
  * the PR to be merged.
  */
 export class AutoMerge extends Component {
-  public readonly autoMergeLabel: string;
```
Yeap

iliapolo commented on src/node-package.ts:
```diff
@@ -716,7 +716,7 @@ export class NodePackage extends Component {
   }
 
-  private renderInstallCommand(frozen: boolean) {
+  public renderInstallCommand(frozen: boolean) {
```
It avoid duplicating the exact same logic in the upgrade workflow. Since this method doesn't expose anything inherently private, figured its harmless.

iliapolo commented on src/node-project.ts:
```diff
@@ -180,15 +178,6 @@ export interface NodeProjectOptions extends ProjectOptions, NodePackageOptions {
-   *
-   * @default "auto-merge"
-   */
-  readonly mergifyAutoMergeLabel?: string;
```
yes

iliapolo commented on src/upgrade-dependencies.ts:
```diff
@@ -0,0 +1,272 @@
+    // then we define a github workflow if possible
+    if (project.github) {
+
+      const schedule = options.schedule ?? UpgradeDependenciesSchedule.DAILY;
```
Not sure which part you mean here exactly. I've refactored the code since this comment so hopefully its more readable and clean now.

iliapolo commented on src/upgrade-dependencies.ts:
```diff
@@ -0,0 +1,272 @@
+
+      const upgradeAndBuild = {
+        'name': 'Upgrade & Build',
+        'permissions': 'read-all',
```
added

iliapolo commented on src/upgrade-dependencies.ts:
```diff
@@ -0,0 +1,272 @@
+              `${this.setOutput(buildLogOutputName, 'https://github.com/${REPO}/actions/runs/${{ github.run_id }}')}`,
+            ].join('\n'),
+            env: {
+              GITHUB_TOKEN: '${{ secrets.GITHUB_TOKEN }}',
```
Its not, letfovers.

iliapolo commented on package.json:
```diff
@@ -18,7 +18,6 @@
     "build": "/bin/bash ./projen.bash build",
     "test:watch": "/bin/bash ./projen.bash test:watch",
     "test:update": "/bin/bash ./projen.bash test:update",
-    "projen:upgrade": "/bin/bash ./projen.bash projen:upgrade",
```
What's the meaning of a `projen:upgrade` task in projen itself? It uses its own source code.

iliapolo commented:
@eladb Addressed all your comments.

Note that I changed the default `dependabot` to `false` - so you'll see an `upgrade-dependencies.yml` workflow here instead. 

Capturing the outcome after this PR:

- When `dependabot` is `true` - use dependabot.
- When `dependabot` is `false` - use new workflow. This might be a little surprising - should we switch to not use anything at all? 
- When `dependabot` is `undefined` - users get the new workflow.

In all cases users get a projen upgrade workflow like they have now in case a projen secret is configured.

eladb reviewed (approved):
Ship it 🛳️
